### PR TITLE
Terra application menu & header

### PR DESCRIPTION
**Pull request by dkasper-was-taken**
### Summary
Create application header and menu HOC. Still needs testing, and waiting for sub components and modal manager to land. Review basic api and core implementation of ApplicationHeader.jsx and ApplicationMenu.jsx.

Resolves: #30 & #31

Thanks for contributing to Terra. 
@cerner/terra

### CODE REVIEW

File: packages/terra-application-header/src/ApplicationHeader.jsx [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    this.onDiscloseUtilty = this.onDiscloseUtilty.bind(this);
+    this.handleRequestClose = this.handleRequestClose.bind(this);
+    this.getTargetRef = this.getTargetRef.bind(this);
+    this.setContentNode = this.setContentNode.bind(this);
+    this.state = { isUtilityOpen: false };
+  }
+
+  onDiscloseUtilty(utility) {
+    if (utility && !this.state.isUtilityOpen) {
+      this.popupContent = utility;
```
Comment (tbiethman) [inline packages/terra-application-header/src/ApplicationHeader.jsx]: Can you keep the utility component stored in state directly and lose the open flag? From what I can tell you're keeping them in sync at the moment anyway.

File: packages/terra-application-header/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Terra App Delegate
+
+
```
Comment (JakeLaCombe) [inline packages/terra-application-header/README.md]: Looks like an extra new line here.

File: packages/terra-application-header/README.md [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+- [Documentation](https://github.com/cerner/terra-core/tree/master/packages/terra-app-delegate/docs)
+- [LICENSE](#license)
+
+## Getting Started
+
+- Install from [npmjs](https://www.npmjs.com): `npm install terra-app-delegate`
+
+## LICENSE
+
+Copyright 2017 Cerner Innovation, Inc.
```
Comment (JakeLaCombe) [inline packages/terra-application-header/README.md]: Make this 2018

File: packages/terra-application-header/src/ApplicationHeader.jsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames/bind';
+import AppDelegate from 'terra-app-delegate';
+import RoutingStackDelegate from 'terra-navigation-layout/lib/RoutingStackDelegate';
+import ApplicationHeaderLayout from 'terra-application-header-layout';
+// import { ApplicationHeaderUtility } from 'terra-application-utility';
```
Comment (JakeLaCombe) [inline packages/terra-application-header/src/ApplicationHeader.jsx]: Don't forget to remove

File: packages/terra-application-header/examples/index-examples/ApplicationHeaderExample.jsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        toggleMenu: () => {},
+        menuIsOpen: false,
+      }}
+      nameConfig={{
+        title: 'Application Name',
+      }}
+      utilityConfig={{
+        userName: 'Some User',
+        userDetails: 'Details about user',
+        onUtilityChange: (event, key) => { this.stuff(key); },
```
Comment (JakeLaCombe) [inline packages/terra-application-header/examples/index-examples/ApplicationHeaderExample.jsx]: What's the stuff method?

File: packages/terra-application-header/src/ApplicationHeader.jsx [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+  /**
+   * Navigational links that will generate list items that will update the path. These paths are matched with react-router to selection.
++  */
+  applicationLinks: PropTypes.arrayOf(PropTypes.shape({
+    id: PropTypes.string,
+    path: PropTypes.string.isRequired,
+    text: PropTypes.string.isRequired,
+  })),
+  /**
+   * Configutation values for the ApplicationName component.
```
Comment (JakeLaCombe) [inline packages/terra-application-header/src/ApplicationHeader.jsx]: *Configuration

File: packages/terra-application-header/src/ApplicationHeader.jsx [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    menuIsOpen: PropTypes.bool,
+    togglePin: PropTypes.bool,
+    menuIsPinned: PropTypes.bool,
+  }).isRequired,
+  /**
+   * Delegate prop that is padded by the NavigationLayout.
+   */
+  routingStackDelegate: RoutingStackDelegate.propType,
+  /**
+   * Configration to be provided to the ApplicationUtility component.
```
Comment (JakeLaCombe) [inline packages/terra-application-header/src/ApplicationHeader.jsx]: *Configuration

File: packages/terra-application-header/src/ApplicationHeader.scss [hunk 8]
```diff
@@ -0,0 +1,26 @@
+:local {
+  .application-header {
+    background-color: rgb(36, 129, 202);;
+    // background-image: durp;
```
Comment (JakeLaCombe) [inline packages/terra-application-header/src/ApplicationHeader.scss]: To be honest, I kinda like this style 😄

File: packages/terra-application-header/src/ApplicationHeader.scss [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    border: 0.07143rem solid rgba(255, 255, 255, 0.17);
+    height: 40px;
```
Comment (JakeLaCombe) [inline packages/terra-application-header/src/ApplicationHeader.scss]: Having a set height like this for the application header restricts it's responsiveness and flexibility. If you do need this, can you do a min-height instead?

File: packages/terra-application-header/src/ApplicationHeader.scss [hunk 10]
```diff
@@ -0,0 +1,26 @@
+:local {
```
Comment (JakeLaCombe) [inline packages/terra-application-header/src/ApplicationHeader.scss]: Should the colors in this file be themeable?

**Comment by JakeLaCombe** [comment]
Out of curiousity, sometimes I see CSS classes written, and other times you used inline styles. In what instances did you decide to use a class, and in what other instances did you decide inline? Consumers ask me about this standard all the time, and I have trouble giving a good answer for what our conventions are.

File: packages/terra-application-header/README.md [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# Terra App Delegate
```
Comment (bjankord) [inline packages/terra-application-header/README.md]: This file is for terra app delegate but is listed as the readme for terra-application-header? 😕

File: packages/terra-application-header/README.md [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+
+[![NPM version](http://img.shields.io/npm/v/terra-app-delegate.svg)](https://www.npmjs.org/package/terra-app-delegate)
+[![Build Status](https://travis-ci.org/cerner/terra-core.svg?branch=master)](https://travis-ci.org/cerner/terra-core)
+
+The AppDelegate is an object that provides a set of functions used to facilitate communcation between Containers and
```
Comment (bjankord) [inline packages/terra-application-header/README.md]: Update description to: `The ApplicationHeader is a HOC that combines config and menu subcomponents, which are then placed within the header layout.`

File: packages/terra-application-header/README.md [hunk 13]
```diff
@@ -0,0 +1,26 @@
+# Terra App Delegate
+
+
+[![NPM version](http://img.shields.io/npm/v/terra-app-delegate.svg)](https://www.npmjs.org/package/terra-app-delegate)
```
Comment (bjankord) [inline packages/terra-application-header/README.md]: Need to update terra-app-delegate to terra-application-header in these 2 urls.

File: packages/terra-application-header/README.md [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+[![Build Status](https://travis-ci.org/cerner/terra-core.svg?branch=master)](https://travis-ci.org/cerner/terra-core)
+
+The AppDelegate is an object that provides a set of functions used to facilitate communcation between Containers and
+their consuming Containers/Applications.
+
+- [Getting Started](#getting-started)
+- [Documentation](https://github.com/cerner/terra-core/tree/master/packages/terra-app-delegate/docs)
```
Comment (bjankord) [inline packages/terra-application-header/README.md]: Need to update terra-app-delegate to terra-application-header in this url.

File: packages/terra-application-header/README.md [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+- [LICENSE](#license)
+
+## Getting Started
+
+- Install from [npmjs](https://www.npmjs.com): `npm install terra-app-delegate`
```
Comment (bjankord) [inline packages/terra-application-header/README.md]: Need to update terra-app-delegate to terra-application-header in the install instructions.

File: packages/terra-application-header/package.json [hunk 16]
```diff
@@ -0,0 +1,66 @@
+{
+  "name": "terra-application-links",
```
Comment (bjankord) [inline packages/terra-application-header/package.json]: This file is located at `packages/terra-application-header/package.json`, though it looks like its using `terra-application-links` in places it should be using `terra-application-header`.

The name key of this package should be updated to `terra-application-header`.

File: packages/terra-application-header/package.json [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+  "main": "lib/ApplicationLinks.js",
```
Comment (bjankord) [inline packages/terra-application-header/package.json]: This should be updated to `lib/terra-application-header`.

File: packages/terra-application-header/package.json [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+  "version": "0.0.0",
+  "description": "{insert description}",
```
Comment (bjankord) [inline packages/terra-application-header/package.json]: Update description to: The ApplicationHeader is a HOC that combines config and menu subcomponents, which are then placed within the header layout.

File: packages/terra-application-header/package.json [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/cerner/terra-framework.git"
+  },
+  "keywords": [
+    "Cerner",
+    "Terra",
+    "Framework",
+    "terra-application-links",
```
Comment (bjankord) [inline packages/terra-application-header/package.json]: Update to `terra-application-header`

File: packages/terra-application-header/package.json [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    "ApplicationLinks",
```
Comment (bjankord) [inline packages/terra-application-header/package.json]: Update to `ApplicationHeader`

File: packages/terra-application-header/src/ApplicationHeader.scss (hunk 10 above)
Comment (bjankord) [reply packages/terra-application-header/src/ApplicationHeader.scss]: Yes, we'll want them to be themeable for the legacy app. Looks like all the colors in this file are now using CSS custom properties.

File: packages/terra-application-header/README.md [hunk 21]
```diff
@@ -0,0 +1,25 @@
+# Terra Application Header
+
+
+[![NPM version](http://img.shields.io/npm/v/terra-application-header.svg)](https://www.npmjs.org/package/terra-application-header)
+[![Build Status](https://travis-ci.org/cerner/terra-framwork.svg?branch=master)](https://travis-ci.org/cerner/terra-work)
```
Comment (bjankord) [inline packages/terra-application-header/README.md]: Missing an `e` here: `terra-framwork.svg` => `terra-framework.svg`
Missing the word `frame` here: `travis-ci.org/cerner/terra-work` => `travis-ci.org/cerner/terra-framework`

File: packages/terra-application-header/src/ApplicationHeader.jsx (hunk 4 above)
Comment (emilyrohrbough) [reply packages/terra-application-header/src/ApplicationHeader.jsx]: I think this is waiting to consumer those lower-level components

File: packages/terra-application-header/src/ApplicationHeader.jsx [hunk 22]
```diff
@@ -0,0 +1,175 @@
...
++  */
+  applicationLinks: PropTypes.arrayOf(PropTypes.shape({
+    id: PropTypes.string,
+    path: PropTypes.string.isRequired,
+    text: PropTypes.string.isRequired,
+  })),
+  /**
+   * Configuration values for the ApplicationName component.
+   */
+  nameConfig: PropTypes.shape({
```
Comment (emilyrohrbough) [inline packages/terra-application-header/src/ApplicationHeader.jsx]: Is there a reason we didn't split these into two separate props?

File: packages/terra-application-header/src/ApplicationHeader.jsx [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+    menuIsPinned: PropTypes.bool,
+  }).isRequired,
+  /**
+   * Delegate prop that is padded by the NavigationLayout.
+   */
+  routingStackDelegate: RoutingStackDelegate.propType,
+  /**
+   * Configuration to be provided to the ApplicationUtility component.
+   */
+  utilityConfig: PropTypes.shape({
```
Comment (emilyrohrbough) [inline packages/terra-application-header/src/ApplicationHeader.jsx]: Same with these, why not create their own props? Seems to leave more room for proptype error

File: packages/terra-application-header/examples/index-examples/ApplicationHeaderExample.jsx (hunk 5 above)
Comment (bjankord) [reply packages/terra-application-header/examples/index-examples/ApplicationHeaderExample.jsx]: Same question.

File: packages/terra-application-header/src/ApplicationHeader.jsx (hunk 4 above)
Comment (bjankord) [reply packages/terra-application-header/src/ApplicationHeader.jsx]: Are these ready to be uncommented?

Comment (dkasper-was-taken) [reply packages/terra-application-header/src/ApplicationHeader.jsx]: Name and Links can be, but Utility is still pending.

**Comment by dkasper-was-taken** [comment]
Closing, as ApplicationMenu and ApplicationHeader are being moved into Application.

**Pull request closed without merging**